### PR TITLE
Use goto_functiont::parameter_identifiers and remove goto_functiont::type

### PR DESCRIPTION
**Pull request by tautschnig**
Having two function type definitions (one in the symbol table and one in
goto_functiont) is just prone to become inconsistent. Instead, always look at
the symbol table when we do need type information. If all we care about are the
identifiers we can still use goto_functiont::parameter_identifiers.

~I'm also marking this RFC, @kroening in particular: it's quite painful to maintain consistency between `parameter_identifiers` and `code_typet::parameters()` - do we really need `parameter_identifiers`, or shouldn't we instead always use the information stored in the type?~

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
Also marking this do-not-review: I should try to break this up into a series of smaller commits that prepare the final change.

**Comment by kroening** [comment]
@tautschnig For a long list of reasons, it's best to get rid of the identifiers in `code_typet::parameters()` and only use the ones in `goto_functiont`.
I am happy to get rid of the signature in `goto_functiont`, for the reason that you are giving.

**Comment by peterschrammel** [comment]
On a side note: I've seen an issue with function pointer types recently: i.e. a function pointer variable has the original return type whereas the symbol table has return type void for the pointed function after the remove_return pass. Type consistency validations fail due to that.

**Comment by kroening** [comment]
On the side note: One cleanup item on my list is to change the way remove_returns works: the idea is to keep the return type as is. This removes the need to keep a copy of the original return type.

**Review by owen-mc-diffblue — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e8d7597a88a06cd5dc84ebb95a89c9122a5733d2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102934571
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

**Comment by kroening** [comment]
I would leave the note about `parameter_identifiers` around for a while, say until `codet::parametert::get_identifier()` is gone.

**Comment by tautschnig** [comment]
> I would leave the note about `parameter_identifiers` around for a while, say until `codet::parametert::get_identifier()` is gone.

Done, and updated - assuming that was the comment you were after.

**Comment by kroening** [comment]
Apologies for iterating on this PR.

The PR goes beyond the parameter identifiers, and also touches the "is hidden" feature. The question of whether or not the execution of a functions should be hidden is a property of the body (it's triggered via a label in the body), and not a property of the function signature.

It should therefore be in goto_functiont, and not the type.

**Comment by martin-cs** [comment]
@tautschnig Do you want this reviewed as is?

**Comment by tautschnig** [comment]
> Apologies for iterating on this PR.
> 
> The PR goes beyond the parameter identifiers, and also touches the "is hidden" feature. The question of whether or not the execution of a functions should be hidden is a property of the body (it's triggered via a label in the body), and not a property of the function signature.
> 
> It should therefore be in goto_functiont, and not the type.

That all makes sense, but really means I should first revert #4312. I'll create a PR to that effect.

**Comment by tautschnig** [comment]
The "is hidden" PR is now posted as #5575.

**Comment by tautschnig** [comment]
@kroening @martin-cs This is now rebased on top of the changes implementing "is hidden" as requested.

**Review by kroening — APPROVED** [review]

File: src/goto-programs/goto_function.h [hunk 1]
```diff
@@ -16,29 +16,23 @@ Date: May 2018
...
-  DEPRECATED(SINCE(2019, 2, 16, "Get the type from the symbol table instead"))
-  code_typet type;
-
   typedef std::vector<irep_idt> parameter_identifierst;
-
   /// The identifiers of the parameters of this function
   ///
   /// Note: This is now the preferred way of getting the identifiers of the
-  /// parameters. The identifiers in the type will go away.
+  /// parameters. The identifiers in the type are now gone.
```
Comment (martin-cs) [inline src/goto-programs/goto_function.h]: :smile: nice!

**Review by martin-cs — APPROVED** [review]
Seems good.

File: src/goto-programs/goto_function.h (hunk 1 above)
Comment (kroening) [reply src/goto-programs/goto_function.h]: I am afraid they are still there. "The type" here refers to the type in the symbol table.

Comment (tautschnig) [reply src/goto-programs/goto_function.h]: Err, true. Change to the comment reverted/dropped.

**Pull request merged**